### PR TITLE
Fix manual conflict resolution for binary files

### LINKED ISSUE #8059: Binary conflict resolve doesn't work
## Description

When I pull repository from remote, there is conflict for an image. When I resolve the conflict with accept the remote branch in GitHub Desktop, and then commit and push, the result isn't expected, it's  still the image in my local, not the image of remote.
image

## Version

* GitHub Desktop: 2.1.0

* Operating system: Windows 10

## Steps to Reproduce

1. commit changes for 1.png
2. pull from remote which has changes for 1.png
3. show the conflict dialog for 1.png
4. chose "use the modified file from origin/xxx"
5. commit merge, and push

### Expected Behavior

the image is the remote one

### Actual Behavior

it's still the local one, didn't change

## Additional Information

Even for banch merge, such as merge master into branch1, the image conflict resolve doesn't work either.

### PR DESCRIPTION
**Pull request by kuychaco**
## Overview

Supersedes #8060 
**Closes #8059 **
Potentially addresses #7166 (needs verification. my guess is that it doesn't, though https://github.com/desktop/desktop/issues/8059#issuecomment-517673659 indicates that it might)

## Description

In certain merge conflict cases, the file under conflict does not have merge markers. Examples include if a file has been deleted on one branch and modified on another, or if the file is a binary file and has been modified on both branches. In Desktop parlance, we call these "Manual conflicts".

For manually resolving binary files, previously if the user selected changes from the `theirs` branch (the branch being merged into the current branch) as part of the merge, they would be unpleasantly surprised to find that the change from the current branch (`ours`) were used in the merge. This was an oversight that never worked correctly (as opposed to a regression). We simply used `git add` to stage the conflicted file, which committed the working directory contents (the version of the file on the current branch). 

With this PR, we ensure that we use the correct version of binary files under conflict by checking out the proper version before staging (`git checkout --ours/theirs -- filePath`). 

Tests were added for choosing both the `theirs` and `ours` versions of conflicted binary files during a merge, which should help protect against future regressions. 

Manual testing steps for merging:

1. download [repo-binary-file.zip](https://github.com/desktop/desktop/files/3494892/repo-binary-file.zip)
1. check out `master` and verify that you're on commit `bbb019c` (or `reset --hard` to it)
1. check out `branch` and verify that you're on commit `0359ae5` (or `reset --hard` to it)
1. merge `master` into `branch`
1. click "Resolve" and choose "Use modified file from master" 
1. click "Commit merge"
1. view the `cute_rodent.jpeg` file and verify that it has rotated left by 90 degrees
1. now redo merge... `git reset --hard 0359ae5` to make it so that we're back to where we started
1. merge `master` into `branch` again
1. click "Resolve" and choose "Use modified file from branch" 
1. click "Commit merge"
1. view the `cute_rodent.jpeg` file and verify that image is now upside-down

Manual testing steps for rebasing:

1. download [repo-binary-file.zip](https://github.com/desktop/desktop/files/3494892/repo-binary-file.zip)
1. check out `master` and verify that you're on commit `bbb019c` (or `reset --hard` to it)
1. check out `branch` and verify that you're on commit `0359ae5` (or `reset --hard` to it)
1. rebase `branch` onto `master`
1. click "Resolve" and choose "Use modified file from master" 
1. click "Continue rebase"
1. view the `cute_rodent.jpeg` file and verify that it has rotated left by 90 degrees
1. now redo rebase... `git reset --hard 0359ae5`  to make it so that we're back to where we started
1. rebase `branch` onto `master` again
1. click "Resolve" and choose "Use modified file from branch" 
1. click "Continue rebase"
1. view the `cute_rodent.jpeg` file and verify that image is now upside-down

## Release notes

Notes: Fixes manual conflict resolution for binary files

### CODE REVIEW

File: app/src/lib/git/core.ts [hunk 1]
```diff
@@ -347,3 +347,18 @@ export async function gitNetworkArguments(
 export function parseCommitSHA(result: IGitResult): string {
   return result.stdout.split(']')[0].split(' ')[1]
 }
+
+/**
+ * Run a Git command and return whether the exit code indicated success
+ *
+ * This defers to the default error handling infrastructure inside if an error
+ * is encountered.
```
Comment (probablycorey) [inline app/src/lib/git/core.ts]: > default error handling infrastructure inside

I wasn't sure what this meant. Does that mean inside this method, or inside something else?

File: app/test/helpers/repositories.ts [hunk 2]
```diff
@@ -225,7 +218,7 @@ export async function setupConflictedRepoWithMultipleFiles(): Promise<
 
   await makeCommit(repo, thirdCommit)
 
-  await FSE.writeFile(filePaths[4], 'touch')
+  await FSE.writeFile(Path.join(repo.path, 'dog'), 'touch')
```
Comment (probablycorey) [inline app/test/helpers/repositories.ts]: 🐶> :cat:

**Comment by kuychaco** [comment]
@niik I incorporated your great feedback. Always nice to leave the code a bit cleaner than when we found it 😄 

Mind re-reviewing?

File: app/src/lib/git/stage.ts [hunk 3]
```diff
@@ -24,47 +24,50 @@ export async function stageManualConflictResolution(
...
+          ? 'theirs'
+          : 'ours'
+      const checkoutCompleted = await git(
+        ['checkout', `--${choiceFlag}`, '--', file.path],
         repository.path,
-        'addConflictedFile'
-      )).exitCode
+        'checkoutConflictedFile'
+      )
+      if (checkoutCompleted) {
```
Comment (niik) [inline app/src/lib/git/stage.ts]: I belive we can drop this check now as well given that the previous `checkout` invocation will throw an exception if it exits with a non-zero exit code.

Comment (kuychaco) [reply app/src/lib/git/stage.ts]: ah good catch!

**Comment by kuychaco** [comment]
@niik addressed your feedback! Mind re-re-reviewing? 😄 

@tierninho nice catch on that bug. It's a sneaky one! I'm going to open a separate issue for it because it's distinct from what's being addressed here. I verified that it's on `development` so it was pre-existing and this PR didn't introduce it. I did some code spelunking and the underlying issue has to do with how we fetch information about the rebase state from the .git folder in `getRebaseInternalState`. I'll share more details in the issue.

**Pull request merged**